website/docs/tutorial/02-making-improvements.md

```diff
@@ -0,0 +1,501 @@
+
+Sure, it's not a _lot_ of wasted space. But every time a user invokes your contract, they will need to pay for the contract's Wasm bytecode to be deserialized from blockchain storage, loaded into the Stellar runtime, and executed. Over the lifetime of your contract and the blockchain, it all adds up!
+
+You can tell Rust that you know what you're doing here to get rid of this waste.
+
+```rust
+pub fn guess(env: &Env, a_number: u64) -> bool {
+    a_number == unsafe { env.storage().instance().get(THE_NUMBER).unwrap_unchecked() }
+}
+```
+
+Or, if you want to clean things up a little and add some comments about why it's ok to use `unsafe` (a good idea!), you could do this:
+
+```rust
+pub fn guess(env: &Env, a_number: u64) -> bool {
+    a_number == Self::number(env)
+}
+
+/// readonly function to get the current number
+fn number(env: &Env) -> u64 {
+    // We can unwrap because the number is set in the constructor
+    // and then only reset by the admin
+    unsafe { env.storage().instance().get(THE_NUMBER).unwrap_unchecked() }
+}
+```
+
+## Your Complete Updated Contract
+
+Here's what your `lib.rs` should look like now:
+
+```rust
+#![no_std]
+use soroban_sdk::{contract, contractimpl, symbol_short, Address, BytesN, Env, Symbol};
+
+#[contract]
+pub struct GuessTheNumber;
+
+const THE_NUMBER: Symbol = symbol_short!("n");
+pub const ADMIN_KEY: &Symbol = &symbol_short!("ADMIN");
+
+#[contractimpl]
+impl GuessTheNumber {
+    pub fn __constructor(env: &Env, admin: &Address) {
+        Self::set_admin(env, admin);
+        Self::set_random_number(env);
+    }
+
+    /// Update the number. Only callable by admin.
+    pub fn reset(env: &Env) {
+        Self::require_admin(env);
+        Self::set_random_number(env);
+    }
+
+    /// Guess a number between 1 and 10
+    pub fn guess(env: &Env, a_number: u64) -> bool {
+        a_number == Self::number(env)
+    }
+
+    /// Private helper function to generate and store a new random number
+    fn set_random_number(env: &Env) {
+        let new_number: u64 = env.prng().gen_range(1..=10);
+        env.storage().instance().set(&THE_NUMBER, &new_number);
+    }
+
+    /// readonly function to get the current number
+    fn number(env: &Env) -> u64 {
+        // We can unwrap because the number is set in the constructor
+        // and then only reset by the admin
+        unsafe { env.storage().instance().get(THE_NUMBER).unwrap_unchecked() }
+    }
+
+    /// Upgrade the contract to new wasm. Only callable by admin.
+    pub fn upgrade(env: &Env, new_wasm_hash: BytesN<32>) {
+        Self::require_admin(env);
+        env.deployer().update_current_contract_wasm(new_wasm_hash);
+    }
+
+    /// Get current admin
+    pub fn admin(env: &Env) -> Option<Address> {
+        env.storage().instance().get(ADMIN_KEY)
+    }
+
+    /// Set a new admin. Only callable by admin.
```

Comment:
The comment is slightly different than the implementation. It can only set the admin once and the initial callee is not authorised actually. I think we should add that in the constructor.